### ISSUE TITLE
Notifications: indicate progress notification if hidden in notification center

### ISSUE THREAD

**Issue by tamuratak** [microsoft/vscode#90274]
Version: 1.42.0
Commit: ae08d5460b5a45169385ff3fd44208f431992451
Date: 2020-02-06T10:51:33.119Z
Electron: 6.1.6
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Darwin x64 17.7.0

Steps to Reproduce:

1. Enable ` Format On Save`.
1. Execute the following extension example.
2. Edit a text file and save it.
3. Cancel the notification`Save Participants`.
4. We cannot even save the text document. 

This issue is a little bit different from #90273. What I want to say here is simple. If `provideDocumentFormattingEdits` throws `Error` after  the notification `Save Participants` canceled, we can not save files anymore. The files remain dirty.

```typescript
import * as vscode from 'vscode';
import { DocumentFormattingEditProvider } from 'vscode';

class Formatter implements DocumentFormattingEditProvider {
	provideDocumentFormattingEdits(document: vscode.TextDocument): vscode.ProviderResult<vscode.TextEdit[]> {
		return new Promise((resolve) => {
			setTimeout(() => {
				throw new Error('always error')
			}, 5000)
		})
	}
}

export function activate(context: vscode.ExtensionContext) {
	console.log('Congratulations, your extension "helloworld-sample" is now active!');

	let disposable = vscode.commands.registerCommand('extension.helloWorld', () => {
		vscode.window.showInformationMessage('Hello World!');
	});

	vscode.languages.registerDocumentFormattingEditProvider(
		{ scheme: 'file' },
		new Formatter()
	)
	context.subscriptions.push(disposable);
}

export function deactivate() { }
```

**Comment by jrieken** [comment]
Hm... This should have been fixed with https://github.com/microsoft/vscode/issues/89654 already. I will investigate

**Comment by jrieken** [comment]
I cannot reproduce that.. Once, I have pressed 'cancel' the file is being saved but since the progress message only shows after 3 seconds (when dirty, 5sec when not dirty) it might appear as not saving.

**Comment by tamuratak** [comment]
I am sorry. I thought pressing `Esc` was the same as pressing `Cancel` on the notification.

The actual steps are

1. Enable ` Format On Save`.
1. Execute the following extension example.
2. Edit a text file and save it.
3. **Press `Esc`** when the notification`Save Participants` appears.
4. We cannot even save the text document.

**Comment by jrieken** [comment]
Ok, I understand now. Yeah, this is a little unfortunate because `Esc` hides the notification but doesn't cancel it. It is then still alive and can be restored by clicking its status bar icon. 

![Feb-10-2020 11-07-32](https://user-images.githubusercontent.com/1794099/74140569-9493b400-4bf5-11ea-9f60-fcb53c876e85.gif)

Adding @bpasero for better UX here. Maybe don't hide on `Esc` when showing progress and show progress in the status bar icon when hiding the real notification?

**Comment by bpasero** [comment]
@misolori nicht need your idea here. I would like to preserve that ESC closes notification toasts because anything else would be quite unpredictable for a user IMHO. 

However giving an indication about a notification that is having progress seems like a good idea to me. Essentially some indicator at the notification bell location that something is going on?

**Comment by misolori** [comment]
The simplest approach would be adding a dot to the bell, though the dot is tied to a "notification" or "activity" so it's not to off:

![image](https://user-images.githubusercontent.com/35271042/74181510-4f23b680-4bf6-11ea-98dd-e7dbdb235a01.png)

<img width="50%" src="https://user-images.githubusercontent.com/35271042/74181521-564ac480-4bf6-11ea-931a-4145538e7348.png">

Adding any smaller detail becomes a bit harder since the icon is so small already:

<img width="50%" src="https://user-images.githubusercontent.com/35271042/74181806-c2c5c380-4bf6-11ea-83df-99115f21504c.png">

**Comment by bpasero** [comment]
@misolori hm, how about an actual animation down there, similar to the sync icon for git spinning when you pull and push?

**Comment by misolori** [comment]
@bpasero what if re-used the same line animation in the sidebar and placed it below the notification bell? 

![Kapture 2020-02-14 at 9 17 50](https://user-images.githubusercontent.com/35271042/74552813-e9dc0800-4f0a-11ea-81cd-8217a5901265.gif)

**Comment by bpasero** [comment]
@misolori yeah that works for me, is there some code I can copy from?

**Comment by misolori** [comment]
@bpasero let me see if I can create this via CSS (this was a simple animation in Figma).

**Comment by misolori** [comment]
@bpasero I was able to use the `:before` pseudo element to create this and animate it via css, so you can add this animation class to the icon when in that state.

https://codepen.io/miguelsolorio/pen/bGdEOvB

**Comment by gediminasz** [comment]
I've recently encountered this issue (when using Python Black formatter with [External Formatters](https://marketplace.visualstudio.com/items?itemName=SteefH.external-formatters) extension and saving a file containing syntax errors) and would like to share my two cents. I would either expect the file to be saved no matter what (with an error notification that the formatter failed of course) or the save process to be cancelled immediately. In other words, if the formatter failed and nothing is running, what is there to cancel?

P.S. Sorry if my comment is irrelevant, love what you guys are doing, autoformatting has changed my life, peace! 😃

**Comment by bpasero** [comment]
@misolori upon second thought, I wonder if a better approach would be to reuse the status location for progress that we already have so that we do not introduce another progress indicator. I mean we have this one:

![Screenshot 2020-02-15 at 14 24 23](https://user-images.githubusercontent.com/900690/74588691-f6da2380-4ffe-11ea-8add-8367e22ea974.png)

So we could do this: if a progress notification is running but not visible, we show it in the status bar at that location keeping the label up to date with what the notification provides. Clicking that entry in the status bar could reveal the notifications center and it would hide automatically once the notification item is visible or done.

**Comment by bpasero** [comment]
Hm, upon second thought, we have 2 cases to support:
* user hides notification toasts (e.g. by pressing ESC key)
* user clicks on the X of the notification that shows progress

I am now thinking that for the first case, we can try something as outlined in https://github.com/microsoft/vscode/issues/90274#issuecomment-586383647 (which also seems to be what Azure Portal is doing). 

However for the second case, the notification is literally gone, though the operation is still running. In that case I would like to indicate to the user that the operation is still ongoing, e.g. like so:

![Kapture 2020-02-17 at 11 44 16](https://user-images.githubusercontent.com/900690/74647139-6b39d180-517b-11ea-99e7-c59937e9132f.gif)

**Comment by stevencl** [comment]
One option would be that clicking the X for a notification about an incomplete operation only hides the toast and decrements the unread notification counter. If the operation is complete, clicking X would dismiss the whole notification. In the long running operation example you could click X to dismiss the toast, decrement the unread notifications counter and then show the animated bar underneath the bell icon. When the user clicks the bell icon they see the notification for the long running operation and they can see that it is still in progress. Once the operation is complete, clicking X would then completely dismiss that notification.

This would mean that there would be no way to ever dismiss notifications about long running operations before they are complete. If that was necessary, we could add an explicit dismiss all command in the notification hub.

**Comment by bpasero** [comment]
Yeah I like this suggestion. This would mean there is a subtle change for the X button depending on the kind of notification:
* `X` for a notification: removes the notification
* `X` for a progress notification: moves to notification center

In theory we could also change the icon for a progress notification to indicate this difference, e.g. like a minimize icon.

I suggest we bring this up in this weeks UX call.

**Comment by dzintars** [comment]
Looks like a similar issue. Saw this first time when made some changes in SVG.TS file.
Reloading window helps for some time.

![image](https://user-images.githubusercontent.com/547420/74873522-07283080-5368-11ea-8876-88a1ab70a14a.png)

**Comment by bpasero** [comment]
@dzintars please report as separate issue, maybe to the extension author

**Comment by bpasero** [comment]
Current UX thinking which I will push to `master`:
* you cannot close a notification in progress (we may want to revisit this for toasts)
* if neither toast nor notifications center is visible, you get a spinning circle in the status bar

![Kapture 2020-02-21 at 15 56 48](https://user-images.githubusercontent.com/900690/75044996-cf81cb80-54c2-11ea-892e-ecaa006ae8dd.gif)

**Comment by bpasero** [comment]
**Verification:**
It is easiest to have some local command for crafting progress notification and then verify https://github.com/microsoft/vscode/issues/90274#issuecomment-589688491 via:

```
class ProgressAction extends Action {

	static readonly ID = 'workbench.action.ProgressAction';
	static readonly LABEL = nls.localize('ProgressAction', "ProgressAction");

	static counter = 0;

	private counter = 0;

	constructor(
		id: string,
		label: string,
		@IProgressService private readonly progressService: IProgressService
	) {
		super(id, label);

		ToggleFullScreenAction.counter++;
	}

	run(): Promise<void> {
		this.progressService.withProgress({
			location: ProgressLocation.Notification,
			title: `[${ToggleFullScreenAction.counter}] Some long running operation`,
			cancellable: true
		}, progress => {

			return timeout(20000);
		});

		return Promise.resolve();
	}
}
```
